### ISSUE TITLE
DefaultJWTEncryptionAndDecryptionService ignores KeyId (kid) Header

### ISSUE THREAD

**Issue by leonard84** [mitreid-connect/OpenID-Connect-Java-Spring-Server#1047]
In `org.mitre.jwt.encryption.service.impl.DefaultJWTEncryptionAndDecryptionService#decryptJwt` the value of the `kid` header is ignored, only the defaultDecryptionKeyId is used to lookup the key/decryptor.

```
JWEDecrypter decrypter = decrypters.get(getDefaultDecryptionKeyId());
```

Shouldn't the code look more like this

```
JWEDecrypter decrypter = null;
String keyID = jwt.getHeader().getKeyID();
if (keyID != null) {
    decrypter = decrypters.get(keyID);
}

decrypter = decrypter == null ? decrypters.get(getDefaultDecryptionKeyId()) : decrypter;
```

**Comment by jricher** [comment]
Yes, that is more correct. I don't think we've had many people deploy with multiple keys, or use the encryption service very much in production.

**Comment by leonard84** [comment]
If you want I can put it in a pull request. 

Btw: is there any way to indicate which key from the jwks should be used by clients? If we want to do a smooth key-rollover the server would need a way to indicate to the client which key it should use. The [OpenID Connect Discovery](https://openid.net/specs/openid-connect-discovery-1_0.html) does not specify anything in this regard, and [JWKS](https://tools.ietf.org/html/rfc7517) doesn't do it either. 

There is actually another problem, if a client specifies more than one key, as it is recommended by the spec

> If the Client signs requests to the Server, it contains the signing key(s) the Server uses to validate signatures from the Client. The JWK Set MAY also contain the Client's encryption keys(s), which are used by the Server to encrypt responses to the Client. When both signing and encryption keys are made available, a use (Key Use) parameter value is REQUIRED for all keys in the referenced JWK Set to indicate each key's intended usage. Although some algorithms allow the same key to be used for both signatures and encryption, doing so is NOT RECOMMENDED, as it is less secure.

`DefaultJWTEncryptionAndDecryptionService` and `DefaultJWTSigningAndValidationService`  ignore the key use parameter in a key set.
Since the methods `getDefaultEncryptionKeyId`, `getDefaultDecryptionKeyId` do not filter for key use as well and will return null if the keyset contains both a sign and encryption key.

```
public String getDefaultEncryptionKeyId() { // getDefaultDecryptionKeyId works similar
		if (defaultEncryptionKeyId != null) {
			return defaultEncryptionKeyId;
		} else if (keys.size() == 1) {
			// if there's only one key in the map, it's the default
			return keys.keySet().iterator().next();
		} else {
			return null;
		}
	}
```

`org.mitre.jwt.signer.service.impl.DefaultJWTSigningAndValidationService#validateSignature` ignores the `kid` as well and just brute-forces the validation which will fill the logs with unnecessary error-level logs.
